Add ScaledBrightnessARGBConverter

tinevez opened the pull request:
This is to enable in BDV a better display control for label images

This converter allows for better controling the visibility of ARGB segmentation. The scale value is used to alter the brightness of colors, which is better suited than altering R G and B in bulk.

This could be used in the BDV, to finely control the visibility of a segmentation overlay in the shape of an ARGB image.

tinevez committed:
Add ScaledBrightnessARGBConverter

This is to enable in BDV a better display control for label images

This converter allows for better controling the visibility of ARGB
segmentation. The scale value is used to alter the brightness of
colors, which is better suited than altering R G and B in bulk.

This could be used in the BDV, to finely control the visibility of a
segmentation overlay in the shape of an ARGB image.

tpietzsch commented:
If possible, I would like to avoid the dependency on `java.awt.Color`.

We could just inline the `Color.HSBtoRGB` and `Color.RGBtoHSB` implementations.
This would also avoid to have the (non-threadsafe) `float[] hsb` array.
But I strongly suspect that full conversion to/from HSB is not necessary to change the brightness.
I didn't look into the math (https://en.wikipedia.org/wiki/HSL_and_HSV#HSV_to_RGB etc) in detail, but from looking over it, I think there should be shortcuts. Do you want to dig into that? (Otherwise, I can also dig a bit deeper when I find time)

tpietzsch commented:
I couldn't resist... I think it should be more or less:

v := max(r,g,b)
m := min(1, v/scale) - v

and then

r' := min(255, r+m)
g' := min(255, g+m)
b' := min(255, b+m)

Could you try whether that works?

tinevez committed:
Rewrite the RGB to HSB routine.

This code is adapted from https://www.cs.rit.edu/~ncs/color/t_convert.html
with feedback from Claude AI.
It improves the previous version by getting rid of the Color dependency
and by being thread safe (removing the need for a float[] field).

tinevez commented:
Done! 
Tested with the BDV overlaying a segmentation results as a label image. The scale properly alters the visibility of the segmentation overlay without touching the teint.

tinevez committed:
Test simple RGB scaling.

tinevez commented:
Here is what I have tested, with a label image overlay in the BDV, in 'Sum' mode, and adjusting the scale to 255, 122, and 0.

## The current implementation

https://github.com/imglib/imglib2/blob/master/src/main/java/net/imglib2/display/ScaledARGBConverter.java#L102

<img width="1112" height="912" alt="imglib2original1" src="https://github.com/user-attachments/assets/3670a8ed-a2b6-43b4-be53-bec6ed8ef512" />
<img width="1112" height="912" alt="imglib2original2" src="https://github.com/user-attachments/assets/365ff723-06b3-4dbe-bbbf-f2a1ce242343" />
<img width="1112" height="912" alt="imglib2original3" src="https://github.com/user-attachments/assets/1598f3bd-6cdd-4062-a348-2d5fc705e98a" />

## Scaling brightness via HSB conversion

https://github.com/imglib/imglib2/blob/183efc225cf3442bfe070671c1499afb2c42ed9a/src/main/java/net/imglib2/display/ScaledBrightnessARGBConverter.java#L47

<img width="1086" height="832" alt="HDBconverter1" src="https://github.com/user-attachments/assets/125abdf5-172a-4ba5-8ee6-853418f2b939" />
<img width="1086" height="832" alt="HDBconverter2" src="https://github.com/user-attachments/assets/6b7429bf-12f5-45ba-89e5-7a74c3eedfbf" />
<img width="1086" height="832" alt="HDBconverter3" src="https://github.com/user-attachments/assets/c476514a-78fe-4fa9-90a4-345cd7dae856" />

## Simple RGB scaling

https://github.com/imglib/imglib2/blob/argb-converter/src/main/java/net/imglib2/display/ScaledBrightnessARGBConverter.java#L47

<img width="1112" height="912" alt="simpleRGBscaling1" src="https://github.com/user-attachments/assets/95151253-8958-440e-bbd4-d47b700f3372" />
<img width="1112" height="912" alt="simpleRGBscaling2" src="https://github.com/user-attachments/assets/93fc17c1-0419-452b-bb31-40d5401aa194" />
<img width="1112" height="912" alt="simpleRGBscaling3" src="https://github.com/user-attachments/assets/b5f017b9-78ec-4f46-91ce-af864bf696d5" />

## Conclusion

The simple RGB scaling gives perceptually similar results and is considerably simpler. I suggest we stick to it.

tpietzsch commented:
Could you try what I suggested above:
(input color: r,g,b)
v := max(r,g,b)
m := min(1, v/scale) - v
r' := min(255, r+m)
g' := min(255, g+m)
b' := min(255, b+m)
(output color: r,g,b)

I think this should be *exactly* the same as via HSB conversion.
(I just took the "convert-to-HSB-then-modify-S-then-convert-to-RGB" sequence and simplify the equations).

tinevez commented:
I had tried and it does not seem to work. The display is unsensitive to `scale` and the colors are wrong. Here is what I implemented

```java
	private static final int getScaledColor( final int color, final double scale )
	{
		final int r = ARGBType.red( color );
		final int g = ARGBType.green( color );
		final int b = ARGBType.blue( color );
		final int v = Math.max( r, Math.max( g, b ) );
		final double m = Math.min( 1., v / scale ) - v;

		final int nr = ( int ) Math.min( 255, r + m );
		final int ng = ( int ) Math.min( 255, g + m );
		final int nb = ( int ) Math.min( 255, b + m );

		return 0xff000000 | ( nr << 16 ) | ( ng << 8 ) | nb;
	}
```

<img width="1112" height="912" alt="Screenshot 2026-01-07 at 15 36 29" src="https://github.com/user-attachments/assets/e00c3656-6974-4ecc-bc36-8044caeae6d8" />

tpietzsch commented:
I looked at the HSB/RGB math again, and surprise: it comes out exactly to your "simple RGB scaling".
If I add +0.5 for rounding, e.g. `final int nr = ( int ) Math.min( 255, r /scale + 0.5 );` it also comes out numerically the same as going through HSB. Cool, I learned something :-)

However, then we can achieve the same result using the existing  `ScaledARGBConverter` and just setting min/max accordingly, that is,
`scaledARGBConverter.setMin(0);`
and
`scaledARGBConverter.setMax(255.0/scale);`

This will not slot into the BDV UI as easily, but I think anyway that this is a mis-use of the min/max sliders. (I'll comment more on the BDV PR)

imagejan commented:
> (I'll comment more on the BDV PR)

For context, here's the link, so others reading this don't have to repeat the search I just did 😉: https://github.com/bigdataviewer/bigdataviewer-core/pull/214
